src/abi/anchor.rs

```diff
@@ -17,3 +29,343 @@ pub fn discriminator(namespace: &'static str, name: &str) -> Vec<u8> {
+            continue;
+        }
+
+        let func = &ns.functions[*func_no];
+        let tags = idl_docs(func);
+
+        let accounts = match &func.mutability {
+            Mutability::Pure(_) => {
+                vec![]
+            }
+            Mutability::View(_) => {
+                vec![IdlAccountItem::IdlAccount(IdlAccount {
+                    name: "data_account".to_string(),
+                    is_mut: false,
+                    is_signer: false,
+                    is_optional: Some(false),
+                    docs: None,
+                    pda: None,
+                    relations: vec![],
+                })]
+            }
+            _ => {
+                vec![IdlAccountItem::IdlAccount(IdlAccount {
+                    name: "data_account".to_string(),
+                    is_mut: true,
+                    is_signer: false,
+                    is_optional: Some(false),
+                    docs: None,
+                    pda: None,
+                    relations: vec![],
+                })]
+            }
+        };
+
+        let mut args: Vec<IdlField> = Vec::with_capacity(func.params.len());
+        for (item_no, item) in func.params.iter().enumerate() {
+            let name = if item.id.is_none() || item.id.as_ref().unwrap().name.is_empty() {
+                format!("arg_{}", item_no)
+            } else {
+                item.id.as_ref().unwrap().name.clone()
+            };
+            args.push(IdlField {
+                name,
+                docs: None,
+                ty: type_manager.convert(&item.ty),
+            });
+        }
+
+        let returns = if func.returns.is_empty() {
+            None
+        } else if func.returns.len() == 1 {
+            Some(type_manager.convert(&func.returns[0].ty))
+        } else {
+            Some(type_manager.build_struct_for_return(func))
+        };
+
+        instructions.push(IdlInstruction {
+            name: if func.is_constructor() {
+                "new".to_string()
+            } else {
+                func.name.clone()
+            },
+            docs: tags,
+            accounts,
+            args,
+            returns,
+        });
+    }
+
+    instructions
+}
+
+/// This struct accounts all the user defined types used in the contract that need to be present
+/// in the IDL 'types' field.
+struct TypeManager<'a> {
+    namespace: &'a Namespace,
+    added_types: HashSet<Type>,
+    types: Vec<IdlTypeDefinition>,
+}
+
+impl TypeManager<'_> {
+    fn new(ns: &Namespace) -> TypeManager {
+        TypeManager {
+            namespace: ns,
+            added_types: HashSet::new(),
+            types: Vec::new(),
+        }
+    }
+
+    /// Retrieve all the custom types to fill the IDL struct
+    fn move_custom_types(&mut self) -> Vec<IdlTypeDefinition> {
+        let mut custom_types: Vec<IdlTypeDefinition> = Vec::new();
+        std::mem::swap(&mut custom_types, &mut self.types);
+        custom_types
+    }
```

Comment:
Could this take `self` rather than `&mut self` and then return `self.types`?

Reply:
No. Rustc does not allow that.

Reply:
This confuses me as well, please consume `self` here. But would you use this function instead of simply accessing `type_manager.types`?

Reply:
I just tried it and it works fine.

Reply:
The function probably needs a better name, or just removed completely.